### DISCUSSION TITLE
Reports: Possible Caching Bug?

### DISCUSSION THREAD

**Discussion by martinum4** [netbox-community/netbox#6693] [Q&A] [answered]
Hey,
i made a report where i messed up the test definition at first.
After a while i got it working and it runs fine, however, when i try to access the reports page i get a server error: `A report must contain at least one test method. Python version: 3.7.3 NetBox version: 2.11.6`

Surprisingly my report runs fine when initiated directly via the  `manage.py runreport`-Command, after this it is loaded in the Web-UI and i can rerun it from there.
This behaviour is reproducable every time i completely purge my browsercache.

Am I doing something wrong?

My Report-Code (Sorry if it looks ugly or ineffective, I'm only getting started with python):

```
import re
from dcim.choices import DeviceStatusChoices
from dcim.models import Device,Site,Location
from extras.reports import Report

from taggit.managers import _TaggableManager

class Powerlist(Report):
	description = "Locate USV-/PDU-IPs and raise error if no primary ipv4 is set"
	def test_powerdevice_list(self):
		for device in Device.objects.filter(status=DeviceStatusChoices.STATUS_ACTIVE):
			if str(device.device_role.name) == ("Active Power Management"):
				if str(device.primary_ip4) == "None":
					self.log_failure(device, "No IP for device in rack {}, {}, {}".format(device.rack,device.location,device.site))
				else:
					self.log_success(device, "https://{} \n Rack: {}, {}, {}".format(str(device.primary_ip4).partition('/')[0],device.rack,device.location,device.site))
			else:
				continue
```
Kind Regards

Martin

**Comment by candlerb** [comment, marked as answer]
Do you have any other `.py` files in ` /opt/netbox/netbox/reports/`, other than `__init__.py` and the report you showed?  I am wondering if the error refers to some other report.

I am on netbox 2.11.7 (python 3.6) and if I drop your report into the reports directory, it works first time from the GUI.

**Reply by candlerb** [reply]
Try this patch to see if the error message gives you a better clue.

```
diff --git a/netbox/extras/reports.py b/netbox/extras/reports.py
index 64fbffb46..508d58af6 100644
--- a/netbox/extras/reports.py
+++ b/netbox/extras/reports.py
@@ -139,7 +139,7 @@ class Report(object):
                     ('log', []),
                 ])
         if not test_methods:
-            raise Exception("A report must contain at least one test method.")
+            raise Exception(f"A report must contain at least one test method ({self.full_name}).")
         self.test_methods = test_methods

     @property
```

Another thought: when running the report from the command line, were you running it as `root`?  What if you try running as the `netbox` user?  Have you checked if any root-owned `.pyc` files have been left around in `/opt/netbox/netbox/reports/__pycache__/` ?  The web interface would be unable to delete those.

**Reply by martinum4** [reply]
I have other reports too, but these have never caused problems before, the problems started to arise when i started scripting the previously posted report, and the error doesn't appear when i delete the report from this folder.
However there is a folder named `__pycache__` inside the reports folder, should i delete this?

I translated the text portions from german to english so that other people are not too much confused by the language barrier, could characters like ü,ä.ö and ß be a problem?
However i only got one of those in `self.log_failure(device, "Keine IP für Gerät in Rack {}, {}, {}".format(device.rack,device.location,device.site))`

I'll try to update my netbox instance next week, maybe this solves the error.

**Reply by martinum4** [reply]
> Try this patch to see if the error message gives you a better clue.
> 
> ```
>+            raise Exception(f"A report must contain at least one test method ({self.full_name}).")

Restarting netbox after applying this patch seems to have solved my issue for now, i will write again when/if it arises again, thanks for your help so far :)

**Reply by candlerb** [reply]
> could characters like ü,ä.ö and ß be a problem?

There shouldn't be, but if you can reproduce it reliably you can raise it as a bug.

The patch wouldn't have changed anything except the error message given, so if the problem has gone away, it's for some other reason.
